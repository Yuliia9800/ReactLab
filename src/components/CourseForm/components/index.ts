export { default as AddAuthor } from './AddAuthor/AddAuthor';
export { default as Authors } from './Authors/Authors';
export { default as CourseAuthor } from './CourseAuthor/CourseAuthor';
export { default as Duration } from './Duration/Duration';
