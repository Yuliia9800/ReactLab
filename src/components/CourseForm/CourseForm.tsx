import React, { useEffect, useMemo, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate, useParams } from 'react-router-dom';
import { v4 as uuidv4 } from 'uuid';

import { Button, Input, Textarea } from 'common';
import {
	CREATE_COURSE_BUTTON_TEXT,
	DESCRIPTION_INPUT_LABEL,
	DESCRIPTION_INPUT_PLACEHOLDER,
	TITLE_INPUT_LABEL,
	TITLE_INPUT_PLACEHOLDER,
	UPDATE_COURSE_BUTTON_TEXT,
} from 'constant';
import { addCourse, updateCourse } from 'store/courses';
import { Author } from 'types';
import { AppDispatch, RootState } from 'store';
import { addAuthor } from 'store/authors';
import { AddAuthor, Duration, Authors, CourseAuthor } from './components';

function CourseForm() {
	const navigate = useNavigate();
	const dispatch = useDispatch<AppDispatch>();
	const { courseId } = useParams();

	const courses = useSelector((state: RootState) => state.courses);
	const courseInfo = courses.find((course) => course.id === courseId);
	const authors = useSelector((state: RootState) => state.authors);

	const [courseAuthorIds, setCourseAuthorIds] = useState([]);
	const [titleValue, setTitleValue] = useState('');
	const [descriptionValue, setDescriptionValue] = useState('');
	const [durationValue, setDurationValue] = useState(0);
	const availableAuthors = useMemo(
		() => authors.filter((author) => !courseAuthorIds.includes(author.id)),
		[authors, courseAuthorIds]
	);

	const courseAuthors = useMemo(
		() => authors.filter((author) => courseAuthorIds.includes(author.id)),
		[authors, courseAuthorIds]
	);

	useEffect(() => {
		if (courseId && courseInfo) {
			setTitleValue(courseInfo.title);
			setCourseAuthorIds(courseInfo.authors);
			setDescriptionValue(courseInfo.description);
			setDurationValue(courseInfo.duration);
		}
	}, [courseId, courseInfo, authors]);

	const handleTitleChange = (event) => {
		setTitleValue(event.target.value);
	};

	const handleDescriptionChange = (event) => {
		setDescriptionValue(event.target.value);
	};

	const handleDurationChange = (event) => {
		setDurationValue(event.target.value);
	};

	const handleCreateAuthor = (author: Author) => {
		dispatch(addAuthor(author));
	};

	const handleAddAuthorToCourse = (author) => {
		setCourseAuthorIds((prev) => [...prev, author.id]);
	};

	const handleDeleteAuthorFromCourse = (author) => {
		setCourseAuthorIds((prev) => prev.filter((id) => id !== author.id));
	};

	const handleCreate = () => {
		if (durationValue <= 0) {
			alert('Duration should be more than 0 minute');
			return;
		}

		if (
			!titleValue ||
			!descriptionValue ||
			!setCourseAuthorIds.length ||
			!durationValue
		) {
			alert('Please, fill in all fields');
			return;
		}

		const newCourse = {
			id: courseInfo?.id || uuidv4(),
			title: titleValue,
			description: descriptionValue,
			creationDate: new Date().toString(),
			duration: Number(durationValue),
			authors: courseAuthorIds,
		};

		courseId
			? dispatch(updateCourse(newCourse))
			: dispatch(addCourse(newCourse));
		navigate('/courses');
	};

	return (
		<>
			<div className='m-5 flex flex-col justify-between space-y-5 p-5 shadow-xl'>
				<div className='flex justify-between'>
					<Input
						labelText={TITLE_INPUT_LABEL}
						placeholderText={TITLE_INPUT_PLACEHOLDER}
						className='input-sm max-w-sm'
						value={titleValue}
						onChange={handleTitleChange}
					/>

					<Button
						buttonText={
							courseId ? UPDATE_COURSE_BUTTON_TEXT : CREATE_COURSE_BUTTON_TEXT
						}
						className='btn-secondary self-end'
						onClick={handleCreate}
					/>
				</div>

				<Textarea
					value={descriptionValue}
					labelText={DESCRIPTION_INPUT_LABEL}
					onChange={handleDescriptionChange}
					placeholderText={DESCRIPTION_INPUT_PLACEHOLDER}
				/>

				<div className='flex gap-16 p-5 '>
					<div className='flex-1 space-y-8 '>
						<AddAuthor createAuthor={handleCreateAuthor} />
						<Duration
							durationValue={durationValue}
							handleDurationChange={handleDurationChange}
						/>
					</div>

					<div className='flex-1 space-y-8'>
						<Authors
							authors={availableAuthors}
							onClick={handleAddAuthorToCourse}
						/>
						<CourseAuthor
							authors={courseAuthors}
							onClick={handleDeleteAuthorFromCourse}
						/>
					</div>
				</div>
			</div>
		</>
	);
}

export default CourseForm;
