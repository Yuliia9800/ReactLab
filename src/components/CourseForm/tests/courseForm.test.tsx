import React from 'react';
import { renderWithProviders, screen } from 'testing';
import CourseForm from '../CourseForm';

test('should match snapshot', () => {
	const { container } = renderWithProviders(<CourseForm />);

	expect(container).toMatchSnapshot();
});

test('should show authors lists (all and course authors)', () => {
	renderWithProviders(<CourseForm />);

	expect(screen.getByText('Authors')).toBeInTheDocument();
	expect(screen.getByText('Course authors')).toBeInTheDocument();
});
