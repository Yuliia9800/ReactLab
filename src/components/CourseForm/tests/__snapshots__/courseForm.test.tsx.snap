// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div>
  <div
    class="m-5 flex flex-col justify-between space-y-5 p-5 shadow-xl"
  >
    <div
      class="flex justify-between"
    >
      <div
        class="form-control w-full"
      >
        <label
          class="label text-xs font-medium text-gray-700"
        >
          <span
            class="label-text"
          >
            Title
          </span>
        </label>
        <input
          class="input-primary input input-md input-sm max-w-sm"
          placeholder="Enter title..."
          type="text"
          value=""
        />
      </div>
      <button
        class="btn-sm btn px-5 btn-secondary self-end"
        type="button"
      >
        Create course
      </button>
    </div>
    <div
      class="form-control w-full"
    >
      <label
        class="label"
      >
        <span
          class="label-text"
        >
          Description
        </span>
      </label>
      <textarea
        class="textarea-primary textarea p-5"
        placeholder="Enter description..."
      />
    </div>
    <div
      class="flex gap-16 p-5 "
    >
      <div
        class="flex-1 space-y-8 "
      >
        <div
          class="flex flex-col space-y-4 "
        >
          <h3
            class="self-center font-semibold"
          >
            Add author
          </h3>
          <div
            class="form-control w-full"
          >
            <label
              class="label text-xs font-medium text-gray-700"
            >
              <span
                class="label-text"
              >
                Add author name
              </span>
            </label>
            <input
              class="input-primary input input-md input-sm w-full"
              placeholder="Enter author name..."
              type="text"
              value=""
            />
          </div>
          <button
            class="btn-sm btn px-5 btn-outline self-center"
            data-testid="createAuthor"
            type="button"
          >
            Add Author
          </button>
        </div>
        <div
          class="flex flex-col space-y-2"
        >
          <h3
            class="self-center font-semibold"
          >
            Duration
          </h3>
          <div
            class="form-control w-full"
          >
            <label
              class="label text-xs font-medium text-gray-700"
            >
              <span
                class="label-text"
              >
                Duration
              </span>
            </label>
            <input
              class="input-primary input input-md input-sm w-full"
              min="0"
              placeholder="Enter duration in minutes..."
              type="number"
              value="0"
            />
          </div>
          <h4
            class="text-lg font-light"
          >
            Duration:
            <b
              class="text-xl font-bold"
            >
               
              00:00
               
            </b>
            hours
          </h4>
        </div>
      </div>
      <div
        class="flex-1 space-y-8"
      >
        <div
          class="flex flex-col justify-end space-y-4"
        >
          <h3
            class="self-center font-semibold"
          >
            Authors
          </h3>
        </div>
        <div
          class="flex flex-col justify-end space-y-4"
        >
          <h3
            class="self-center font-semibold"
          >
            Course authors
          </h3>
          <p
            class="self-center text-sm font-light"
          >
            Author list is empty
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
